### ISSUE TITLE
[question] compatibility with hashicorp/go-version

### ISSUE THREAD

**Issue by i4ki** [apparentlymart/go-versions#3]
Hey,

First of all, thanks for the amazing library.

I have a question regarding compatibility with [hashicorp/go-version](https://github.com/hashicorp/go-version) library.

We want to migrate to this library for better handling of prereleases.
We will add a `enable-prerelease-exact-match` kind of flag and then allow the Prerelease set in such case but we'd like to keep the same behavior of the old library when the flag is not set.

For the constraint checks using prerelease ordering, I'm using the `versions.MeetingConstraintsExact(spec)` and it's working as expected.

But for the `go-versions` old behavior, I tried to use the snippet below:

```
semver, _ := versions.ParseVersion(version)
spec, _ := constraints.ParseRubyStyleMulti(constraint)

allowed := versions.MeetingConstraints(spec)
return allowed.Has(semver)
```
but it only matches prereleases if they are an exact match (`= operator`) but old library allows for any operator given that a prerelease is requested. 

See the testcases below:
https://go.dev/play/p/b_cvnYfl6xn

Example: In the hashicorp library, if the constraint is `< 1.2.3-beta` then the version `1.2.3-alpha` matches.

So my question is: is it possible to massage the `versions.Set` in a such a way that it'd allow for any operator in the prerelease check?

**Comment by apparentlymart** [comment]
Hi @i4ki! Sorry for the slow response.

Aside from the cosmetic decision of using a similar interpretation of what "ruby-like" constraint syntax looks like, this library is not really designed to be fully compatible with `hashicorp/go-version` behavior.

There is no built-in way to get the effect you want here but I think you could potentially implement it yourself:

In this library I intentionally separated the constraint syntax parser (package `constraints`) from the version set model so that it would be possible to implement other constraint syntaxes or to interpret the built-in constraint syntax in different ways.

To implement the behavior you want I think you could first call `versions.MeetingConstraintsExact` to get the base version set that doesn't yet filter out the prereleases, and then write your own implementation of `versions.MeetingConstraints` (which is itself a wrapper around `versions.MeetingConstraintsExact`) which uses some custom logic written by you to analyze a [`constraints.Spec`](https://pkg.go.dev/github.com/apparentlymart/go-versions@v1.0.2/versions/constraints#Spec) to find all of the leaf `VersionSpec` and collect them all into a separate `versions.List` which will be your own definition of the "requested versions".

After that you can follow the same strategy `versions.MeetingConstraintExact` follows to construct a union of your requested versions and the "exact" versions set. Make sure to refer to v1.0.2 of this library as your starting point because I've just fixed #4 where previously that function was not correctly handling negated exact selections.

**Comment by i4ki** [comment]
Hi @apparentlymart 

Thanks for the detailed answer. Yes, in the meantime I implemented the logic myself, on top of the abstractions exported by your package. 
Thank you very much!
